storage/core/src/main/java/com/hortonworks/streamline/storage/CacheBackedStorageManager.java

```diff
@@ -52,32 +52,50 @@ public void init(Map<String, Object> properties) {
     @Override
     public void add(Storable storable) throws StorageException {
         writer.add(storable);
-        cache.put(storable.getStorableKey(), storable);
+        if (storable.isCacheable()) {
+            cache.put(storable.getStorableKey(), storable);
+        }
     }
 
     @Override
     public <T extends Storable> T remove(StorableKey key) throws StorageException {
-        writer.remove(key);
-        final T oldVal = (T) cache.get(key);
-        cache.remove(key);
-        return oldVal;
+        Storable storable = (Storable) writer.remove(key);
+        if (storable != null && storable.isCacheable()) {
+            storable = cache.get(key);
+            cache.remove(key);
+        }
+        return (T) storable;
     }
 
     @Override
     public void addOrUpdate(Storable storable) throws StorageException {
         writer.addOrUpdate(storable);
-        cache.put(storable.getStorableKey(), storable);
+        if (storable.isCacheable()) {
+            cache.put(storable.getStorableKey(), storable);
+        }
+    }
+
+    @Override
+    public void update(Storable storable) {
+        writer.update(storable);
+        if (storable.isCacheable()) {
+            cache.put(storable.getStorableKey(), storable);
+        }
     }
 
     @Override
     public <T extends Storable> T get(StorableKey key) throws StorageException {
-        return (T) cache.get(key);
+        Storable storable = cache.get(key);
+        if (storable != null && !storable.isCacheable()) {
```

Comment:
In `GuavaCache` implementation we load the values via a Loading cache so by invoking `cache.get(key)` it automatically loads it into cache. So we need to remove it. Will add a comment to make it clear.